### ISSUE TITLE
Looking for a document where the face landmark locations are defined

### ISSUE THREAD

**Issue by fangq** [google/mediapipe#3799]
my student and I are working on a robust registration between face-mesh with other head/brain-landmarks used for neuroimaging, specifically, the [10-20 system](https://en.wikipedia.org/wiki/10%E2%80%9320_system_(EEG)) for EEG.

the comments in a previous issue (#1615) seem to give a good picture how these face landmarks are numbered, but I still don't see a rigorous definition on where these key points are supposed to be from a face-feature perspective, for example, where is landmark 10 is supposed to be located? middle of the forehead? is there a standard or document where I can find face landmark's "actual" geometric definitions?

thanks

**Comment by kuaashish** [comment]
Hi @fangq,
Is it implemented through Python, Javascript or other solutions available in MediaPipe? Could you please provide the complete details to understand the issue better from our end. Thank you!

**Comment by fangq** [comment]
Currently, our initial implementation is based on python, but the question is more generic and independent of implementation.

on a related note - among the 468 face mesh points, are there "dependent" points that are computed/interpolated based on other independently estimated points? if so, what are those and how they are interpolated?

**Comment by rajkundu** [comment]
I'm not sure if this helps at all, but there is a .obj file located [here](https://github.com/google/mediapipe/blob/master/mediapipe/modules/face_geometry/data/canonical_face_model.obj) in the MediaPipe repository. You can open this file in external software to visualize and analyze the face mesh model. For example, using [MeshLab](https://www.meshlab.net/), you can get the index of a vertex in a mesh as shown below:

<img width="1904" alt="image" src="https://user-images.githubusercontent.com/35236205/198916941-a8607662-8274-43b1-9706-101187736bba.png">

To do so:
1) Open MeshLab, which will create a new, empty project.
2) Drag & drop `canoncal_face_model.obj` onto the window, which will import the mesh into this new project.
3) Click the "Get Info" button (yellow circle with an "i" in the icon bar).
4) Press "t" to switch from face-selection mode (default) to vertex-selection mode.
5) Click any vertex to get its index.

I am very much a beginner when it comes to MeshLab, but if you have any questions, I would be happy to try and answer them. I recommend playing around with the vertex display size as shown in my screenshot above, on the right, in the sidebar.

**Comment by fangq** [comment]
@rajkundu thanks for your reply, but I don't think it answers the question I was asking - this still only gives the numbering of points, but not what face feature they are supposed to be pointing to.

just an example - what location on the face that point 10 is supposed to be?

**Comment by rajkundu** [comment]
Sorry; I'm not entirely sure what you mean. To me, it seems like that question is answerable by looking at the [canonical face model UV map](https://github.com/google/mediapipe/blob/a908d668c730da128dfa8d9f6bd25d519d006692/mediapipe/modules/face_geometry/data/canonical_face_model_uv_visualization.png), then seeing where Point 10 is. Or, if you're interested in Point 10's 3D definition, I would look at the .obj file.

Perhaps you're looking for an _anatomical_ definition instead?

**Comment by fangq** [comment]
> Perhaps you're looking for an anatomical definition instead?

yes, that is correct. I previously referred it to as "geometric definition", but I think "anatomical definition" is more appropriate.

**Comment by rajkundu** [comment]
Ah, I see. To be honest, I'm not quite sure how these types of things tend to work. I think the question now is whether Google's engineers attached anatomical definitions to any of the the MP face model's points. I mean, to some extent, strict anatomical correspondences may not even be necessary for the model to function as intended.

My guess is as good as yours, but I would imagine the MP face model does not have many, if any, anatomical correspondences. Some of the only easy/robust ones that I can think of are: the corners of the eyes, the tip of the nose, and maybe even the nasion (vertex 168?). In general, I imagine that finding these correspondences is our job; I don't think there is an official document on them, though I could certainly be wrong.

**Comment by fangq** [comment]
yes, the only anatomically-clear landmark relevant to my project is nasion (pt168), other than that, I have been computing other 10-20 anchor points empirically by extrapolating 10, 234, 227, 457 and 447, which worked reasonably well, but I was hoping if all landmarks were anatomically defined, then I would get more assets to improve the mapping.

**Comment by andrechen** [comment]
It always puzzle me to find out the the landmark id. So I injected [.obj file](https://github.com/andrechen/mediapipe/tree/master/mediapipe/modules/face_geometry/data) with material section. And use a simple material file to bind the uv visualization map. It makes the obj could be opened in Mesh viewer with vertex id...
![canonical_face_model_3D_viewer](https://user-images.githubusercontent.com/2261137/198922587-4bc3689d-1822-4122-b233-907196069fa7.png)

or you can import canonical face model to blender, then find interpolation weights to define your customized landmarks...
![canonical_face_model_blender](https://user-images.githubusercontent.com/2261137/198923077-21676610-0c9f-428b-ac52-029ef25d6544.png)

**Comment by fangq** [comment]
@kuaashish, thanks for following up. We are still looking for official documentation on these landmarks. My student had spent the past two  semesters trying to fit these facial landmarks to anatomically defined head landmarks using a large library, but if knowing more on their definitions will still help for refine this registration.

**Comment by kuaashish** [comment]
@fangq,

Thanks for the confirmation, For now, We do not have any anatomical naming scheme for each of the 468 landmark. However, We can mark this as feature request and share internally based on discussion, We will prioritise the request.

**Comment by rezinghost** [comment]
i'm searching for definitions too, and i just found this [keypoints](https://github.com/ninuannvarghese/TensorFlowJS/blob/613e163bc9b21fc9691af578c084aee615fc95fa/facemesh/src/keypoints.ts) the orders of landmark indexes are mirrored e.g. in rightEyeUpper0 and leftEyeUpper0. But still lack of that on other facial parts like check or forehead.
